lms/static/scripts/frontend_apps/components/SubmitGradeForm.tsx

```diff
@@ -103,6 +107,10 @@ export default function SubmitGradeForm({
   // Make sure instructors are notified if there's a risk to lose unsaved data
   useWarnOnPageUnload(hasUnsavedChanges);
 
+  useEffect(() => {
+    onUnsavedChanges?.(hasUnsavedChanges);
+  }, [onUnsavedChanges, hasUnsavedChanges]);
```

Comment:
> It would be preferable to call this from the place that sets `hasUnsavedChanges`, so that the parent can re-render at the same time. You might want to wrap the code that calls `setHasUnsavedChanges` and invokes this callback in a helper to ensure both are always called together.

`hasUnsavedChanges` is a memoized state derived from other pieces of local state. It does not have a setter per se, that's why I had to wrap it in a `useEffect` here.

I will check if this can be redesigned, but considering the amount of things involved in considering the form "dirty", this was a nice solution.

> Also, there is a possible confusion here that re-rendering this component with a change to the value of the `onUnsavedChanges` callback will invoke the callback, even if the value of `hasUnsavedChanges` has not changed. Callbacks from child components don't usually behave this way.

Hhm, good point.

Reply:
Ok, I came up with something that covers both points and I think it should not be very hard to evolve once we introduce the comment field.